discord/ext/commands/core.py

```diff
@@ -476,31 +476,42 @@ async def _actual_conversion(self, ctx, converter, argument, param):
             raise BadArgument('Converting to "{}" failed for parameter "{}".'.format(name, param.name)) from exc
 
     async def do_conversion(self, ctx, converter, argument, param):
-        try:
-            origin = converter.__origin__
-        except AttributeError:
-            pass
-        else:
-            if origin is typing.Union:
-                errors = []
-                _NoneType = type(None)
-                for conv in converter.__args__:
-                    # if we got to this part in the code, then the previous conversions have failed
-                    # so we should just undo the view, return the default, and allow parsing to continue
-                    # with the other parameters
-                    if conv is _NoneType and param.kind != param.VAR_POSITIONAL:
-                        ctx.view.undo()
-                        return None if param.default is param.empty else param.default
-
-                    try:
-                        value = await self._actual_conversion(ctx, conv, argument, param)
-                    except CommandError as exc:
-                        errors.append(exc)
-                    else:
+        origin = typing.get_origin(converter)
+
+        if origin is typing.Union:
+            errors = []
+            _NoneType = type(None)
+            for conv in typing.get_args(converter):
+                # if we got to this part in the code, then the previous conversions have failed
+                # so we should just undo the view, return the default, and allow parsing to continue
+                # with the other parameters
+                if conv is _NoneType and param.kind != param.VAR_POSITIONAL:
+                    ctx.view.undo()
+                    return None if param.default is param.empty else param.default
+
+                try:
+                    value = await self.do_conversion(ctx, conv, argument, param)
+                except CommandError as exc:
+                    errors.append(exc)
+                else:
+                    return value
+
+            # if we're here, then we failed all the converters
+            raise BadUnionArgument(param, typing.get_args(converter), errors)
+
+        if origin is typing.Literal:
+            errors = []
+            for literal in self._flattened_typing_literal_args(converter):
+                try:
+                    value = await self._actual_conversion(ctx, type(literal), argument, param)
+                except CommandError as exc:
+                    errors.append(exc)
+                else:
+                    if value == literal:
                         return value
```

Comment:
Only convert to each unique type once.

```suggestion
            conversions = {}
            literal_args = self._flattened_typing_literal_args(converter)
            for literal in literal_args:
                literal_type = type(literal)
                try:
                    value = conversions[literal_type]
                except KeyError:
                    try:
                        value = await self._actual_conversion(ctx, literal_type, argument, param)
                    except CommandError as exc:
                        errors.append(exc)
                        conversions[literal_type] = object()
                        continue
                    else:
                        conversions[literal_type] = value
                    
                if conversions[literal_type] == literal:
                    return value
```